pkg/util/active_groups.go

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+	maxGroupsPerUser  int
+}
+
+func NewActiveGroups(maxGroupsPerUser int) *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+		maxGroupsPerUser:  maxGroupsPerUser,
+	}
+}
+
+// UpdateGroupTimestampForUser function is only guaranteed to update to the
+// timestamp provided even if it is smaller than the existing value
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, now time.Time) {
+	ts := now.UnixNano()
+	ag.mu.RLock()
+	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
+		ag.mu.RUnlock()
+		groupTs.Store(ts)
+		return
+	}
+	ag.mu.RUnlock()
+
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	if groupTimestamps == nil {
+		ag.timestampsPerUser[userID] = map[string]*atomic.Int64{group: atomic.NewInt64(ts)}
+		return
+	}
+
+	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
+		groupTs.Store(ts)
+		return
+	}
+
+	ag.timestampsPerUser[userID][group] = atomic.NewInt64(ts)
+}
+
+func (ag *ActiveGroups) PurgeInactiveGroupsForUser(userID string, deadline int64) []string {
+	ag.mu.RLock()
+	totalGroups := len(ag.timestampsPerUser[userID])
+	if totalGroups == 0 {
+		ag.mu.RUnlock()
+		return nil
+	}
+
+	inactiveGroups := make([]string, 0, totalGroups)
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	for group, ts := range groupTimestamps {
+		if ts.Load() <= deadline {
+			inactiveGroups = append(inactiveGroups, group)
+		}
+	}
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil || len(inactiveGroups) == 0 {
+		return nil
+	}
+
+	// Cleanup inactive groups
+	deletedGroups := make([]string, 0, len(inactiveGroups))
```

Comment:
This was my suggestion to simplify the code. We started with an in-place modification code like this:

https://github.com/grafana/mimir/blob/8492f40ec657ef68b206ad10319571f9839c3fb5/pkg/util/active_user.go#L76-L96

However, that code has some drawbacks: it locks/unlocks multiple times, which is already worse than just allocating a slice, plus it's 3x longer and requires more cognitive load to understand. (It also removes elements by copying the entire rest of the slice instead of just moving one element and making the slice shorter, but that's easy to fix).

Reply:
My comment is just about re-allocating for `deletedGroups` while you could modify `inactiveGroups` in place. No other change. No change to locking or so.